File: web/src/lib/ws.js
import { api } from './api.js';
import {
  appState, messages, liveByThread, agentsByThread, pendingApproval, wsConnected,
  connection, connFailure, lastTurnUsage, jobs, jobsClock, setLive, appendLiveText,
  clearLive, liveFor, setAgents, clearAgents,
} from '../stores/state.js';
import { get } from 'svelte/store';

let ws = null;

// sendFrame writes to whichever socket is currently live, and reports whether it
// went out. It is a plain function rather than a closure rebuilt on every
// connect: the rebuilt version was null until the first connection and could be
// left behind by a stale reconnect, both of which read as "no connection" while
// one was in fact open.
function sendFrame(obj) {
  if (ws && ws.readyState === WebSocket.OPEN) {
    ws.send(JSON.stringify(obj));
    return true;
  }
  return false;
}

// The tools that launch a subagent. The CLI renamed this one (Task → Agent), so
// both are recognised rather than whichever version happens to be installed.
const SUBAGENT_TOOLS = new Set(['Agent', 'Task']);

// viewing reports whether threadId is the thread the user is currently looking
// at. Turn output is stamped with its thread id by the server; we only touch the
// visible `messages`/appState for the open thread, while every thread's live
// stream keeps accumulating in liveByThread regardless of what is on screen.
function viewing(threadId) {
  return threadId && get(appState)?.thread?.id === threadId;
}

// classifyFailure asks the REST side why the socket is down. A 401 means the
// server is alive and refusing this page's token — almost always because the
// server was restarted and issued a new one, leaving tabs opened before it
// holding a dead link.
async function classifyFailure() {
  try {
    const res = await fetch('/api/state', { headers: { Authorization: `Bearer ${api.token}` } });
    if (res.status === 401 || res.status === 403) return 'auth';
    return res.ok ? null : 'down';
  } catch {
    return 'down'; // nothing answering at all
  }
}

export function connectWS() {
  if (ws) return;

  const proto = location.protocol === 'https:' ? 'wss' : 'ws';
  // Every handler below is bound to this exact socket and checks that it is
  // still the current one before touching shared state.
  //
  // Without that, reconnecting corrupts itself. A retry can leave more than one
  // socket alive for a moment, and handlers that read the module-level `ws`
  // act on whichever socket is current rather than their own: a dying socket's
  // error handler closes its replacement, and its close handler sets the
  // reference to null even though the replacement is open and healthy. The
  // socket then keeps delivering frames — the UI carries on updating — while
  // every send fails, because the code believes there is no connection.
  const sock = new WebSocket(`${proto}://${location.host}/ws`, ['ccl-bearer', api.token]);
  ws = sock;
  const current = () => ws === sock;

  sock.addEventListener('open', () => {
    if (!current()) { sock.close(); return; }
    wsConnected.set(true);
    connFailure.set(null);
    // A fresh socket cannot resume a turn that was streaming to a previous
    // connection, so any leftover "streaming" slice is stale. Clear it so the
    // composer never stays locked behind a ghost spinner after a reconnect.
    // Dropping the socket cancels those turns server-side, so their subagents
    // are gone too and must not be left on screen looking alive.
    liveByThread.set({});
    agentsByThread.set({});
  });

  sock.addEventListener('close', () => {
    // A socket that has already been replaced must not report its death as the
    // current connection's, or it clears a live one.
    if (!current()) return;
    wsConnected.set(false);
    ws = null;
    // The WebSocket API never exposes the HTTP status behind a failed upgrade, so
    // a rejected token and an unreachable server arrive here identically. Ask
    // over REST, which does report it: the difference decides both the advice
    // given and whether retrying is worth anything at all.
    classifyFailure().then(kind => {
      connFailure.set(kind);
      // A rejected token will be rejected again for as long as this page lives —
      // it is the one in the address bar. Retrying forever just hides the real
      // problem behind a spinner.
      if (kind !== 'auth') setTimeout(connectWS, 3000);
    });
  });

  sock.addEventListener('error', () => {
    // Close this socket, never "the current one" — they are not always the same.
    sock.close();
  });

  sock.addEventListener('message', (ev) => {
    if (!current()) return; // frames from a socket already replaced
    let msg;
    try { msg = JSON.parse(ev.data); } catch { return; }
    handleMessage(msg);
  });
}

function handleMessage(msg) {
  switch (msg.type) {
    case 'state':
      appState.set(msg.state);
      if (msg.state?.connection) connection.set(msg.state.connection);
      break;

    case 'connection':
      // Live VPN/API health push from the server's background probe.
      if (msg.status) connection.set(msg.status);
      break;

    case 'jobs':
      // Full snapshot of supervised background jobs, so it replaces the list.
      jobs.set(msg.jobs || []);
      if (msg.now) jobsClock.set(msg.now);
      break;

    case 'event':
      handleEvent(msg);
      break;

    case 'turn_end': {
      // Finalize the live assistant text into the message list, but only if the
      // user is viewing that thread; otherwise the server has persisted it and
      // it will load when they reopen the thread. Either way, drop the live slice.
      const tid = msg.threadId;
      const live = liveFor(tid).text;
      if (live && viewing(tid)) {
        messages.update(ms => [
          ...ms,
          { role: 'assistant', content: live, ts: new Date().toISOString() },
        ]);
      }
      clearLive(tid);
      break;
    }

    case 'approval_request':
      pendingApproval.set(msg);
      break;

    case 'error':
      // A turn-level error from the server. With a threadId it belongs to one
      // turn; without one (e.g. no open project) it refers to the send that just
      // failed, so clear the current thread's optimistic streaming state.
      handleTurnError(msg.threadId, msg.message);
      break;
  }
}

function handleEvent(msg) {
  const tid = msg.threadId;
  switch (msg.kind) {
    case 'text':
      appendLiveText(tid, msg.text || '');
      break;

    case 'tool_start':
      setLive(tid, { streaming: true, tool: msg.tool || 'tool' });
      // Launching a subagent is not a line in the transcript: the panel below
      // says who was launched and what became of them, in more detail than a
      // bare tool name ever could.
      if (viewing(tid) && !SUBAGENT_TOOLS.has(msg.tool)) {
        messages.update(ms => [
          ...ms,
          { role: 'tool', content: msg.tool || '', ts: new Date().toISOString(), _transient: true },
        ]);
      }
      break;

    case 'agents':
      // A full snapshot of the turn's subagents, so it replaces rather than merges.
      setAgents(tid, msg.agents, msg.now);
      break;

    case 'system_init':
      // Carry the new session id onto the turn's thread, if it is the open one.
      if (viewing(tid)) {
        appState.update(s => s ? {
          ...s,
          thread: s.thread ? { ...s.thread, sessionId: msg.sessionId } : s.thread,
        } : s);
      }
      break;

    case 'result': {
      // A result ends a reply, not necessarily the turn: a subagent reports back
      // after the model has already answered, and the model answers again on the
      // same stream. Each reply is closed off here — as the server does when it
      // persists them — so two replies do not run together into one paragraph.
      const done = liveFor(tid).text || msg.text || '';
      if (done && viewing(tid)) {
        messages.update(ms => [
          ...ms,
          { role: 'assistant', content: done, ts: new Date().toISOString() },
        ]);
      }
      setLive(tid, { text: '', tool: '' });
      // Cost/context/model/usage are session-global; update regardless of thread.
      appState.update(s => s ? {
        ...s,
        cost: msg.cost ?? s.cost,
        ctxUsed: msg.ctxUsed ?? s.ctxUsed,
        ctxWindow: msg.ctxWindow ?? s.ctxWindow,
        modelActual: msg.modelActual ?? s.modelActual,
        usage: msg.usage ?? s.usage,
      } : s);
      if (msg.turnUsage) lastTurnUsage.set(msg.turnUsage);
      break;
    }

    case 'retry':
      if (viewing(tid)) {
        messages.update(ms => [
          ...ms,
          { role: 'system', content: `Повтор попытки ${msg.attempt}...`, ts: new Date().toISOString() },
        ]);
      }
      break;

    case 'notice':
      if (viewing(tid)) {
        messages.update(ms => [
          ...ms,
          { role: 'system', content: msg.text || '', ts: new Date().toISOString() },
        ]);
      }
      break;

    case 'rate_limit':
      appState.update(s => {
        if (!s) return s;
        const limits = (s.limits || []).filter(l => l.type !== msg.limitType);
        limits.push({ type: msg.limitType, resetsAt: msg.limitResets, status: msg.limitStatus });
        limits.sort((a, b) => (a.type === 'five_hour' ? 0 : 1) - (b.type === 'five_hour' ? 0 : 1));
        return { ...s, limits };
      });
      break;

    case 'error':
      handleTurnError(tid, msg.error || '');
      break;
  }
}

function handleTurnError(threadId, text) {
  const tid = threadId || get(appState)?.thread?.id;
  if (viewing(tid) && text) {
    messages.update(ms => [
      ...ms,
      { role: 'system', content: `Ошибка: ${text}`, ts: new Date().toISOString(), _error: true },
    ]);
  }
  clearLive(tid);
}

// sendFailureText explains why a message could not be sent. Telling someone to
// reload is worse than useless when the token is the problem: it lives in this
// page's URL, so a reload sends the same rejected token again.
function sendFailureText(kind) {
  if (kind === 'auth') {
    return 'Сервер отклонил токен этой вкладки — скорее всего он был перезапущен и выдал новый. ' +
      'Обновление страницы не поможет: токен зашит в её адрес. ' +
      'Возьмите свежую ссылку из терминала, где запущен сервер, и откройте её в новой вкладке.';
  }
  // The overwhelmingly common cause is a server that is restarting — a rebuild
  // takes a few seconds — not one that has gone away. Saying "check it is
  // running" sent people looking for a problem that fixes itself, so the text
  // says what is actually true: the text is kept, and the socket comes back.
  return 'Связь с локальным сервером на секунду пропала — сообщение не отправлено. ' +
    'Текст сохранён в поле ввода: переподключение идёт автоматически, ' +
    'через пару секунд просто нажмите Enter ещё раз.';
}

// sendMessage dispatches a turn and reports whether it actually went out, so
// the composer can keep the text when it did not. Anything else throws away
// what someone just typed — and the moment it happens is a reconnect, when they
// are least likely to still have it.
export function sendMessage(text, attachmentPaths) {
  const threadId = get(appState)?.thread?.id;
  if (!threadId) {
    // Silently doing nothing here is how a message disappears with no
    // explanation at all — the case someone hits right after a restart, before
    // a project is open again.
    messages.update(ms => [
      ...ms,
      {
        role: 'system',
        content: 'Нет открытого проекта — сообщение отправлять некуда. ' +
          'Выберите проект в боковой панели, текст сохранён.',
        ts: new Date().toISOString(),
        _error: true,
      },
    ]);
    return false;
  }
  // Transmit first: if the socket is down the send is dropped, and faking a
  // spinner + user message here would lock the composer behind a turn that was
  // never dispatched. Surface the failure instead of swallowing it.
  if (!sendFrame({ type: 'send', text, attachments: attachmentPaths })) {
    messages.update(ms => [
      ...ms,
      { role: 'system', content: sendFailureText(get(connFailure)), ts: new Date().toISOString(), _error: true },
    ]);
    return false;
  }
  // Bind this turn's live state to the thread it is sent from, so its output
  // renders only there even if the user switches threads mid-turn. The previous
  // turn's subagents go with it — they belong to the exchange above.
  setLive(threadId, { streaming: true, text: '', tool: '' });
  clearAgents(threadId);
  // Add user message to the visible list immediately (current thread == threadId).
  messages.update(ms => [
    ...ms,
    { role: 'user', content: text, attachments: attachmentPaths, ts: new Date().toISOString() },
  ]);
  return true;
}

export function cancelTurn() {
  const threadId = get(appState)?.thread?.id;
  sendFrame({ type: 'cancel', threadId });
  clearLive(threadId);
}

export function sendApproval(id, allow, rememberRule) {
  const msg = { type: 'approval', id, allow };
  if (allow && rememberRule) msg.rememberRule = rememberRule;
  sendFrame(msg);
  pendingApproval.set(null);
}

File: internal/web/ws.go
package web

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"net/http"
	"sync"

	"github.com/gorilla/websocket"

	"github.com/AlexandrKhromov2005/claude-code-linux-ui/internal/core"
)

// handleWS authenticates and upgrades a WebSocket connection. The token is read
// from Sec-WebSocket-Protocol (never the URL), and Origin/Host are validated by
// the upgrader's CheckOrigin and the explicit check below.
func (s *Server) handleWS(w http.ResponseWriter, r *http.Request) {
	if reason := s.allow.checkHostOrigin(r); reason != "" {
		http.Error(w, reason, http.StatusForbidden)
		return
	}
	if !s.validWSToken(r) {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	ws, err := s.upgrader.Upgrade(w, r, nil)
	if err != nil {
		return // upgrader already wrote the error
	}
	c := &wsConn{
		ws:      ws,
		s:       s,
		pending: map[string]chan core.ApprovalDecision{},
		turns:   map[string]*turnHandle{},
	}
	s.setActiveConn(c)
	defer func() {
		s.clearActiveConn(c)
		c.cancelAll()
		c.failPending()
		ws.Close()
	}()
	_ = c.writeJSON(map[string]any{"type": "state", "state": s.state()})
	c.readLoop()
}

// validWSToken reports whether the offered subprotocols include the session
// token alongside the negotiated subprotocol name.
func (s *Server) validWSToken(r *http.Request) bool {
	for _, p := range websocket.Subprotocols(r) {
		if p != wsSubprotocol && tokenEqual(p, s.token) {
			return true
		}
	}
	return false
}

// RequestApproval routes a gated tool call to the connected web client. It
// satisfies core.ApprovalBroker.
func (s *Server) RequestApproval(ctx context.Context, req core.ApprovalRequest) core.ApprovalDecision {
	s.mu.Lock()
	c := s.activeConn
	s.mu.Unlock()
	if c == nil {
		return core.ApprovalDecision{Allow: false, Message: "нет подключённого веб-клиента"}
	}
	return c.requestApproval(ctx, req)
}

// wsConn is one connected web client.
type wsConn struct {
	ws *websocket.Conn
	s  *Server

	writeMu sync.Mutex

	mu      sync.Mutex
	pending map[string]chan core.ApprovalDecision
	// turns holds the in-flight turn for each thread id, so turns in different
	// threads run concurrently and can be cancelled independently. A *turnHandle
	// (not a bare cancel func) so a finishing turn can tell whether it is still
	// the current one for its thread — func values are not comparable, pointers
	// are.
	turns map[string]*turnHandle
}

// turnHandle owns the cancellation of one in-flight turn.
type turnHandle struct {
	cancel context.CancelFunc
}

type inMsg struct {
	Type         string   `json:"type"`
	Text         string   `json:"text"`
	Attachments  []string `json:"attachments"`
	ID           string   `json:"id"`
	Allow        bool     `json:"allow"`
	RememberRule string   `json:"rememberRule"`
	ThreadID     string   `json:"threadId"`
}

func (c *wsConn) writeJSON(v any) error {
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	return c.ws.WriteJSON(v)
}

func (c *wsConn) readLoop() {
	for {
		var m inMsg
		if err := c.ws.ReadJSON(&m); err != nil {
			return
		}
		switch m.Type {
		case "send":
			c.startTurn(m.Text, m.Attachments)
		case "cancel":
			c.cancelTurn(m.ThreadID)
			c.failPending()
		case "approval":
			dec := core.ApprovalDecision{Allow: m.Allow, RememberRule: m.RememberRule}
			if !m.Allow {
				dec.Message = "Отклонено пользователем"
			}
			c.deliver(m.ID, dec)
		}
	}
}

func (c *wsConn) startTurn(text string, attachments []string) {
	ctx, cancel := context.WithCancel(context.Background())
	threadID, ch, err := c.s.app.SendTurn(ctx, text, attachments)
	if err != nil {
		cancel()
		_ = c.writeJSON(map[string]any{"type": "error", "message": err.Error()})
		return
	}
	// Register this turn under its thread, replacing only an earlier turn for the
	// same thread (a re-send). Turns in other threads keep running, which is what
	// lets the user start a task in one project while another is still working.
	h := &turnHandle{cancel: cancel}
	c.registerTurn(threadID, h)
	go func() {
		for ev := range ch {
			m := eventToMsg(ev)
			m["threadId"] = threadID
			// The result event carries this turn's cost; the client shows the
			// running session total, so report the accumulated value instead.
			// Context usage and the effective model ride along on the same event.
			if ev.Kind == core.EvResult {
				m["cost"] = c.s.app.Cost()
				used, win := c.s.app.ContextInfo()
				m["ctxUsed"] = used
				m["ctxWindow"] = win
				m["modelActual"] = c.s.app.ModelActual()
			}
			_ = c.writeJSON(m)
		}
		_ = c.writeJSON(map[string]any{"type": "turn_end", "threadId": threadID})
		c.finishTurn(threadID, h)
		cancel()
	}()
}

// registerTurn records h as the in-flight turn for threadID, cancelling whatever
// turn was running for that same thread before (a re-send).
func (c *wsConn) registerTurn(threadID string, h *turnHandle) {
	c.mu.Lock()
	old := c.turns[threadID]
	c.turns[threadID] = h
	c.mu.Unlock()
	if old != nil {
		old.cancel()
	}
}

// finishTurn clears h as the in-flight turn for threadID, but only if a newer
// re-send has not already replaced it.
func (c *wsConn) finishTurn(threadID string, h *turnHandle) {
	c.mu.Lock()
	if c.turns[threadID] == h {
		delete(c.turns, threadID)
	}
	c.mu.Unlock()
}

// cancelTurn cancels the in-flight turn for one thread, leaving others running.
func (c *wsConn) cancelTurn(threadID string) {
	c.mu.Lock()
	h := c.turns[threadID]
	delete(c.turns, threadID)
	c.mu.Unlock()
	if h != nil {
		h.cancel()
	}
}

// cancelAll cancels every in-flight turn; used when the connection tears down.
func (c *wsConn) cancelAll() {
	c.mu.Lock()
	turns := c.turns
	c.turns = map[string]*turnHandle{}
	c.mu.Unlock()
	for _, h := range turns {
		h.cancel()
	}
}

// requestApproval sends an approval request and blocks until the client answers,
// the turn is cancelled, or the connection drops.
func (c *wsConn) requestApproval(ctx context.Context, req core.ApprovalRequest) core.ApprovalDecision {
	id := randID()
	ch := make(chan core.ApprovalDecision, 1)
	c.mu.Lock()
	c.pending[id] = ch
	c.mu.Unlock()

	msg := map[string]any{
		"type":     "approval_request",
		"id":       id,
		"toolName": req.ToolName,
		"input":    json.RawMessage(req.Input),
		"preview":  core.BuildToolPreview(req.ToolName, req.Input),
	}
	if err := c.writeJSON(msg); err != nil {
		c.discard(id)
		return core.ApprovalDecision{Allow: false, Message: "ошибка отправки запроса"}
	}
	select {
	case dec := <-ch:
		return dec
	case <-ctx.Done():
		c.discard(id)
		return core.ApprovalDecision{Allow: false, Message: "отменено"}
	}
}

func (c *wsConn) deliver(id string, dec core.ApprovalDecision) {
	c.mu.Lock()
	ch := c.pending[id]
	delete(c.pending, id)
	c.mu.Unlock()
	if ch != nil {
		ch <- dec
	}
}

func (c *wsConn) discard(id string) {
	c.mu.Lock()
	delete(c.pending, id)
	c.mu.Unlock()
}

// failPending denies every outstanding approval so blocked permission handlers
// unblock when the turn is cancelled or the client disconnects.
func (c *wsConn) failPending() {
	c.mu.Lock()
	pending := c.pending
	c.pending = map[string]chan core.ApprovalDecision{}
	c.mu.Unlock()
	for _, ch := range pending {
		ch <- core.ApprovalDecision{Allow: false, Message: "соединение закрыто"}
	}
}

// eventToMsg maps a core event to the WebSocket wire shape.
func eventToMsg(ev core.Event) map[string]any {
	m := map[string]any{"type": "event", "kind": eventKind(ev.Kind)}
	if ev.Text != "" {
		m["text"] = ev.Text
	}
	if ev.Tool != "" {
		m["tool"] = ev.Tool
	}
	if ev.Model != "" {
		m["model"] = ev.Model
	}
	if ev.SessionID != "" {
		m["sessionId"] = ev.SessionID
	}
	if ev.CostUSD != 0 {
		m["cost"] = ev.CostUSD
	}
	if ev.Attempt != 0 {
		m["attempt"] = ev.Attempt
	}
	if ev.Err != nil {
		m["error"] = ev.Err.Error()
	}
	if ev.LimitType != "" {
		m["limitType"] = ev.LimitType
		m["limitResets"] = ev.LimitResets
		m["limitStatus"] = ev.LimitStatus
	}
	return m
}

func eventKind(k core.EventKind) string {
	switch k {
	case core.EvText:
		return "text"
	case core.EvToolStart:
		return "tool_start"
	case core.EvSystemInit:
		return "system_init"
	case core.EvResult:
		return "result"
	case core.EvError:
		return "error"
	case core.EvRetry:
		return "retry"
	case core.EvNotice:
		return "notice"
	case core.EvRateLimit:
		return "rate_limit"
	default:
		return "unknown"
	}
}

func randID() string {
	b := make([]byte, 8)
	if _, err := rand.Read(b); err != nil {
		return "id"
	}
	return hex.EncodeToString(b)
}
